Participation via public link not possible

### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### What went wrong, what did you observe?

After updating polls to version 7.0.0 yesterday, participation via public link is no longer possible.
The poll can be viewed, but no voting inputs are visible.

### What did you expect, how polls should behave instead?

Voting was possible before the update, and should be possible now.

### What steps does it need to replay this bug?

1. create a poll
2. share via public link
3. try to vote, without being authenticated


### Installation method

Installed/updated using occ

### Installation type

Updated from previous major version (i.e. 6.x.x to 7.x.x)

### Affected polls version

7.0.0

### Which browser did you use, when experiencing the bug?

- [X] Firefox
- [ ] Chrome
- [ ] Chromium/Chromium based (i.e. Edge)
- [X] Safari
- [ ] Other/Don't know

### Other browser

_No response_

### Add your browser log here

```shell
[Error] [ERROR] Files_PDFViewer: But this does not appear to be a public page – {app: "Files_PDFViewer", level: 0}
	value (files_pdfviewer-public.js:2:11033)
	value (files_pdfviewer-public.js:2:11503)
	(anonyme Funktion) (files_pdfviewer-public.js:2:56822)
[Error] TypeError: null is not an object (evaluating 'o.value')
	(anonyme Funktion) (ownpad-public.js:1:21186)
[Error] Failed to load resource: the server responded with a status of 500 () (app, line 0) (https://cloud.***.de/apps/polls/settings/app?time=1711647540405)
[Error] Failed to load resource: the server responded with a status of 403 () (comments, line 0) (https://cloud.***.de/apps/polls/s/***/comments?time=1711647540713)
```


### Additional client environment information

_No response_

### NC version

Nextcloud 28

### Other Nextcloud version

28.0.4

### PHP engine version

PHP 8.2

### Other PHP version

_No response_

### Database engine

MariaDB

### Database Engine version or other Database

10.3.39

### Which user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other/Don't know

### Add your nextcloud server log here

```json
{
    "reqId": "mZykwILEcvwrJxS969CK",
    "level": 3,
    "time": "2024-03-28T16:34:49+00:00",
    "remoteAddr": "***",
    "user": "***",
    "app": "index",
    "method": "POST",
    "url": "/apps/polls/settings/app",
    "message": "OCA\\Polls\\Model\\Settings\\AppSettings::setBooleanSetting(): Argument #2 ($value) must be of type bool, string given, called in /***/nextcloud/apps/polls/lib/Service/SettingsService.php on line 49 in file '/***/nextcloud/apps/polls/lib/Model/Settings/AppSettings.php' line 181",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4.1 Safari/605.1.15",
    "version": "28.0.4.1",
    "exception": {
        "Exception": "Exception",
        "Message": "OCA\\Polls\\Model\\Settings\\AppSettings::setBooleanSetting(): Argument #2 ($value) must be of type bool, string given, called in /***/nextcloud/apps/polls/lib/Service/SettingsService.php on line 49 in file '/***/nextcloud/apps/polls/lib/Model/Settings/AppSettings.php' line 181",
        "Code": 0,
        "Trace": [
            {
                "file": "/***/nextcloud/lib/private/AppFramework/App.php",
                "line": 184,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->"
            },
            {
                "file": "/***/nextcloud/lib/private/Route/Router.php",
                "line": 315,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::"
            },
            {
                "file": "/***/nextcloud/lib/base.php",
                "line": 1069,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->"
            },
            {
                "file": "/***/nextcloud/index.php",
                "line": 39,
                "function": "handleRequest",
                "class": "OC",
                "type": "::"
            }
        ],
        "File": "/***/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
        "Line": 169,
        "Previous": {
            "Exception": "TypeError",
            "Message": "OCA\\Polls\\Model\\Settings\\AppSettings::setBooleanSetting(): Argument #2 ($value) must be of type bool, string given, called in /***/nextcloud/apps/polls/lib/Service/SettingsService.php on line 49",
            "Code": 0,
            "Trace": [
                {
                    "file": "/***/nextcloud/apps/polls/lib/Service/SettingsService.php",
                    "line": 49,
                    "function": "setBooleanSetting",
                    "class": "OCA\\Polls\\Model\\Settings\\AppSettings",
                    "type": "->"
                },
                {
                    "file": "/***/nextcloud/apps/polls/lib/Controller/SettingsController.php",
                    "line": 60,
                    "function": "writeAppSettings",
                    "class": "OCA\\Polls\\Service\\SettingsService",
                    "type": "->"
                },
                {
                    "file": "/***/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 230,
                    "function": "writeAppSettings",
                    "class": "OCA\\Polls\\Controller\\SettingsController",
                    "type": "->"
                },
                {
                    "file": "/***/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 137,
                    "function": "executeController",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->"
                },
                {
                    "file": "/***/nextcloud/lib/private/AppFramework/App.php",
                    "line": 184,
                    "function": "dispatch",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->"
                },
                {
                    "file": "/mnt/HC_Volume_2665849/nextcloud/lib/private/Route/Router.php",
                    "line": 315,
                    "function": "main",
                    "class": "OC\\AppFramework\\App",
                    "type": "::"
                },
                {
                    "file": "/***/nextcloud/lib/base.php",
                    "line": 1069,
                    "function": "match",
                    "class": "OC\\Route\\Router",
                    "type": "->"
                },
                {
                    "file": "/***/nextcloud/index.php",
                    "line": 39,
                    "function": "handleRequest",
                    "class": "OC",
                    "type": "::"
                }
            ],
            "File": "/***/nextcloud/apps/polls/lib/Model/Settings/AppSettings.php",
            "Line": 181
        },
        "message": "OCA\\Polls\\Model\\Settings\\AppSettings::setBooleanSetting(): Argument #2 ($value) must be of type bool, string given, called in /***/nextcloud/apps/polls/lib/Service/SettingsService.php on line 49 in file '/***/nextcloud/apps/polls/lib/Model/Settings/AppSettings.php' line 181",
        "exception": {},
        "CustomMessage": "OCA\\Polls\\Model\\Settings\\AppSettings::setBooleanSetting(): Argument #2 ($value) must be of type bool, string given, called in /***/nextcloud/apps/polls/lib/Service/SettingsService.php on line 49 in file '/***/nextcloud/apps/polls/lib/Model/Settings/AppSettings.php' line 181"
    }
}

{
    "reqId": "8iPnXzlGoS1kA2I2bbvv",
    "level": 3,
    "time": "2024-03-28T17:52:36+00:00",
    "remoteAddr": "***",
    "user": "--",
    "app": "index",
    "method": "GET",
    "url": "/apps/polls/settings/app?time=1711648356149",
    "message": "Cannot assign null to property OCA\\Polls\\Model\\Group\\Group::$group of type OCP\\IGroup",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4.1 Safari/605.1.15",
    "version": "28.0.4.1",
    "exception": {
        "Exception": "TypeError",
        "Message": "Cannot assign null to property OCA\\Polls\\Model\\Group\\Group::$group of type OCP\\IGroup",
        "Code": 0,
        "Trace": [
            {
                "file": "/***/nextcloud/apps/polls/lib/Model/Group/Group.php",
                "line": 48,
                "function": "setUp",
                "class": "OCA\\Polls\\Model\\Group\\Group",
                "type": "->"
            },
            {
                "file": "/***/nextcloud/apps/polls/lib/Model/Settings/AppSettings.php",
                "line": 208,
                "function": "__construct",
                "class": "OCA\\Polls\\Model\\Group\\Group",
                "type": "->"
            },
            {
                "function": "jsonSerialize",
                "class": "OCA\\Polls\\Model\\Settings\\AppSettings",
                "type": "->"
            },
            {
                "file": "/***/nextcloud/lib/public/AppFramework/Http/JSONResponse.php",
                "line": 71,
                "function": "json_encode"
            },
            {
                "file": "/***/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 182,
                "function": "render",
                "class": "OCP\\AppFramework\\Http\\JSONResponse",
                "type": "->"
            },
            {
                "file": "/***/nextcloud/lib/private/AppFramework/App.php",
                "line": 184,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->"
            },
            {
                "file": "/***/nextcloud/lib/private/Route/Router.php",
                "line": 315,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::"
            },
            {
                "file": "/***/nextcloud/lib/base.php",
                "line": 1069,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->"
            },
            {
                "file": "/***/nextcloud/index.php",
                "line": 39,
                "function": "handleRequest",
                "class": "OC",
                "type": "::"
            }
        ],
        "File": "/***/nextcloud/apps/polls/lib/Model/Group/Group.php",
        "Line": 52,
        "message": "Cannot assign null to property OCA\\Polls\\Model\\Group\\Group::$group of type OCP\\IGroup",
        "exception": {},
        "CustomMessage": "Cannot assign null to property OCA\\Polls\\Model\\Group\\Group::$group of type OCP\\IGroup"
    }
}
```

### Additional environment informations

_No response_

### Configuration report

_No response_

### List of activated Apps

```shell
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/***REMOVED***",
        "dbtype": "mysql",
        "version": "28.0.4.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "maintenance": false,
        "default_language": "de",
        "default_phone_region": "DE",
        "theme": "",
        "loglevel": 0,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "app_install_overwrite": [
            "spreed",
            "ownpad",
            "richdocumentscode",
            "whiteboard",
            "mailnotifier"
        ],
        "mail_sendmailmode": "smtp",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mysql.utf8mb4": true,
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": true,
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0
        },
        "updater.release.channel": "stable"
    }
}
```


### Nextcloud Signing status

```shell
No errors have been found.
```


### Additional Information

_No response_

--- a/lib/AppConstants.php
+++ b/lib/AppConstants.php
@@ -33,6 +33,8 @@ abstract class AppConstants {
 	/** @var string */
 	public const SESSION_KEY_SHARE_TOKEN = 'ncPollsPublicToken';
 	/** @var string */
+	public const SESSION_KEY_SHARE_TYPE = 'ncPollsShareType';
+	/** @var string */
 	public const CLIENT_ID = 'ncPollsClientId';
 	/** @var string */
 	public const CLIENT_TZ = 'ncPollsClientTimeZone';

--- a/lib/Controller/BasePublicController.php
+++ b/lib/Controller/BasePublicController.php
@@ -27,6 +27,7 @@ namespace OCA\Polls\Controller;
 
 use Closure;
 use OCA\Polls\AppConstants;
+use OCA\Polls\Db\ShareMapper;
 use OCA\Polls\Exceptions\Exception;
 use OCA\Polls\Exceptions\NoUpdatesException;
 use OCA\Polls\Model\Acl;
@@ -46,6 +47,7 @@ class BasePublicController extends Controller {
 		IRequest $request,
 		protected ISession $session,
 		protected Acl $acl,
+		protected ShareMapper $shareMapper,
 	) {
 		parent::__construct($appName, $request);
 	}
@@ -98,6 +100,13 @@ class BasePublicController extends Controller {
 	}
 
 	private function updateSessionToken(string $token): void {
+		if (!$token) {
+			$this->session->remove(AppConstants::SESSION_KEY_SHARE_TOKEN);
+			$this->session->remove(AppConstants::SESSION_KEY_SHARE_TYPE);
+			return;
+		}
+		$share = $this->shareMapper->findByToken($token);
 		$this->session->set(AppConstants::SESSION_KEY_SHARE_TOKEN, $token);
+		$this->session->set(AppConstants::SESSION_KEY_SHARE_TYPE, $share->getType());
 	}
 }

--- a/lib/Controller/PublicController.php
+++ b/lib/Controller/PublicController.php
@@ -26,6 +26,7 @@ declare(strict_types=1);
 namespace OCA\Polls\Controller;
 
 use OCA\Polls\AppConstants;
+use OCA\Polls\Db\ShareMapper;
 use OCA\Polls\Model\Acl;
 use OCA\Polls\Service\CommentService;
 use OCA\Polls\Service\MailService;
@@ -58,6 +59,7 @@ class PublicController extends BasePublicController {
 		IRequest $request,
 		ISession $session,
 		Acl $acl,
+		ShareMapper $shareMapper,
 		private IUserSession $userSession,
 		private CommentService $commentService,
 		private MailService $mailService,
@@ -68,7 +70,7 @@ class PublicController extends BasePublicController {
 		private VoteService $voteService,
 		private WatchService $watchService
 	) {
-		parent::__construct($appName, $request, $session, $acl);
+		parent::__construct($appName, $request, $session, $acl, $shareMapper);
 	}
 
 	/**

--- a/lib/Db/Poll.php
+++ b/lib/Db/Poll.php
@@ -31,6 +31,7 @@ use JsonSerializable;
 use OCA\Polls\AppConstants;
 use OCA\Polls\Exceptions\NoDeadLineException;
 use OCA\Polls\Helper\Container;
+use OCP\ISession;
 use OCP\IURLGenerator;
 
 /**
@@ -117,6 +118,7 @@ class Poll extends EntityWithUser implements JsonSerializable {
 	private IURLGenerator $urlGenerator;
 	protected UserMapper $userMapper;
 	private VoteMapper $voteMapper;
+	private ISession $session;
 
 	// schema columns
 	public $id = null;
@@ -170,6 +172,7 @@ class Poll extends EntityWithUser implements JsonSerializable {
 		$this->urlGenerator = Container::queryClass(IURLGenerator::class);
 		$this->userMapper = Container::queryClass(UserMapper::class);
 		$this->voteMapper = Container::queryClass(VoteMapper::class);
+		$this->session = Container::queryClass(ISession::class);
 	}
 
 	/**
@@ -254,6 +257,10 @@ class Poll extends EntityWithUser implements JsonSerializable {
 		}
 		if ($this->getIsCurrentUserLocked() && $this->userRole === self::ROLE_ADMIN) {
 			return self::ROLE_USER;
+		}
+
+		if ($this->session->get(AppConstants::SESSION_KEY_SHARE_TYPE) === Share::TYPE_PUBLIC) {
+			return Share::TYPE_PUBLIC;
 		}
 
 		return $this->userRole;

--- a/lib/Middleware/RequestAttributesMiddleware.php
+++ b/lib/Middleware/RequestAttributesMiddleware.php
@@ -44,6 +44,7 @@ class RequestAttributesMiddleware extends Middleware {
 		if (!$this->hasAttribute($reflectionMethod, PublicPage::class)) {
 			// authenticated session don't use share tokens
 			$this->session->remove(AppConstants::SESSION_KEY_SHARE_TOKEN);
+			$this->session->remove(AppConstants::SESSION_KEY_SHARE_TYPE);
 		}
 
 	}

--- a/lib/Service/CommentService.php
+++ b/lib/Service/CommentService.php
@@ -29,6 +29,7 @@ use OCA\Polls\Db\Comment;
 use OCA\Polls\Db\CommentMapper;
 use OCA\Polls\Event\CommentAddEvent;
 use OCA\Polls\Event\CommentDeleteEvent;
+use OCA\Polls\Exceptions\ForbiddenException;
 use OCA\Polls\Model\Acl;
 use OCP\EventDispatcher\IEventDispatcher;
 
@@ -53,7 +54,11 @@ class CommentService {
 		if ($pollId !== null) {
 			$this->acl->setPollId($pollId);
 		}
-		$this->acl->request(Acl::PERMISSION_COMMENT_ADD);
+		try {
+			$this->acl->request(Acl::PERMISSION_COMMENT_ADD);
+		} catch (ForbiddenException $e) {
+			return [];
+		}
 
 		$comments = $this->commentMapper->findByPoll($this->acl->getPoll()->getId());
 		// treat comments from the same user within 5 minutes as grouped comments

